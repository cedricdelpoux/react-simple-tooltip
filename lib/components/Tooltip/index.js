"use strict"

exports.__esModule = true

var _templateObject = _taggedTemplateLiteralLoose(
    ["\n  0% {\n      opacity: 0;\n  }\n  100% {\n      opacity: 1;\n  }\n"],
    ["\n  0% {\n      opacity: 0;\n  }\n  100% {\n      opacity: 1;\n  }\n"]
  ),
  _templateObject2 = _taggedTemplateLiteralLoose(
    ["\n  animation: ", "ms ", " 0s 1 ", ";\n"],
    ["\n  animation: ", "ms ", " 0s 1 ", ";\n"]
  ),
  _templateObject3 = _taggedTemplateLiteralLoose(
    ["\n  position: absolute;\n  ", ";\n  ", ";\n"],
    ["\n  position: absolute;\n  ", ";\n  ", ";\n"]
  ),
  _templateObject4 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  bottom: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-bottom: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  bottom: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-bottom: ",
      "px;\n",
    ]
  ),
  _templateObject5 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  top: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-top: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  top: 100%;\n  left: 50%;\n  transform: translateX(-50%);\n  margin-top: ",
      "px;\n",
    ]
  ),
  _templateObject6 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  right: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-right: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  right: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-right: ",
      "px;\n",
    ]
  ),
  _templateObject7 = _taggedTemplateLiteralLoose(
    [
      "\n  ",
      ";\n  left: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-left: ",
      "px;\n",
    ],
    [
      "\n  ",
      ";\n  left: 100%;\n  top: 50%;\n  transform: translateY(-50%);\n  margin-left: ",
      "px;\n",
    ]
  )

var _react = require("react")

var _react2 = _interopRequireDefault(_react)

var _propTypes = require("prop-types")

var _propTypes2 = _interopRequireDefault(_propTypes)

var _core = require("@emotion/core")

var _propTypes3 = require("../../utils/propTypes")

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {default: obj}
}

function _objectWithoutProperties(obj, keys) {
  var target = {}
  for (var i in obj) {
    if (keys.indexOf(i) >= 0) continue
    if (!Object.prototype.hasOwnProperty.call(obj, i)) continue
    target[i] = obj[i]
  }
  return target
}

function _taggedTemplateLiteralLoose(strings, raw) {
  strings.raw = raw
  return strings
} /** @jsx jsx */
/* eslint-disable react/prop-types */
/* eslint-disable no-unused-vars */
/* eslint-disable react/react-in-jsx-scope */
/* eslint-disable react/display-name */

var fadeAnimation = (0, _core.keyframes)(_templateObject)

var animation = function animation(props) {
  return (0, _core.css)(
    _templateObject2,
    props.fadeDuration,
    props.fadeEasing,
    fadeAnimation
  )
}

// prettier-ignore
// eslint-disable-next-line no-unused-vars
var Base = function Base(props) {
  return (0, _core.css)(_templateObject3, props.fadeDuration && props.fadeDuration > 0 && animation(props), props.zIndex && "z-index: " + props.zIndex + ";");
};

var Top = function Top(props) {
  return (0, _core.css)(_templateObject4, Base(props), props.offset)
}

var Bottom = function Bottom(props) {
  return (0, _core.css)(_templateObject5, Base(props), props.offset)
}

var Left = function Left(props) {
  return (0, _core.css)(_templateObject6, Base(props), props.offset)
}

var Right = function Right(props) {
  return (0, _core.css)(_templateObject7, Base(props), props.offset)
}

var tooltips = {
  left: function left(_ref) {
    var children = _ref.children,
      props = _objectWithoutProperties(_ref, ["children"])

    return (0, _core.jsx)("div", {css: Left(props)}, children)
  },
  top: function top(_ref2) {
    var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"])

    return (0, _core.jsx)("div", {css: Top(props)}, children)
  },
  right: function right(_ref3) {
    var children = _ref3.children,
      props = _objectWithoutProperties(_ref3, ["children"])

    return (0, _core.jsx)("div", {css: Right(props)}, children)
  },
  bottom: function bottom(_ref4) {
    var children = _ref4.children,
      props = _objectWithoutProperties(_ref4, ["children"])

    return (0, _core.jsx)("div", {css: Bottom(props)}, children)
  },
}

var Tooltip = function Tooltip(_ref5) {
  var children = _ref5.children,
    offset = _ref5.offset,
    open = _ref5.open,
    placement = _ref5.placement,
    zIndex = _ref5.zIndex,
    fadeDuration = _ref5.fadeDuration,
    fadeEasing = _ref5.fadeEasing

  var Component = tooltips[placement] || tooltips.top
  return (
    open &&
    (0, _core.jsx)(
      Component,
      {
        offset: offset,
        zIndex: zIndex,
        fadeDuration: fadeDuration,
        fadeEasing: fadeEasing,
      },
      children
    )
  )
}

Tooltip.propTypes =
  process.env.NODE_ENV !== "production"
    ? {
        children: _propTypes2.default.any.isRequired,
        offset: _propTypes2.default.number,
        open: _propTypes2.default.bool,
        placement: _propTypes2.default.string,
        zIndex: _propTypes2.default.number,
        fadeEasing: _propTypes3.easingPropType,
        fadeDuration: _propTypes2.default.number,
      }
    : {}

exports.default = Tooltip
module.exports = exports["default"]
