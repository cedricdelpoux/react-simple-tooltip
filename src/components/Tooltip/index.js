/** @jsx jsx */
import PropTypes from "prop-types"
import {css, keyframes, jsx} from "@emotion/core"
import {easingPropType} from "../../utils/propTypes"

const fadeAnimation = keyframes`
  0% {
      opacity: 0;
  }
  100% {
      opacity: 1;
  }
`

const animation = props => css`
  animation: ${props.fadeDuration}ms ${props.fadeEasing} 0s 1 ${fadeAnimation};
`

// prettier-ignore
const Base = (props) => css`
  position: absolute;
  ${props.fadeDuration && props.fadeDuration > 0 && animation(props)};
  ${props.zIndex && `z-index: ${props.zIndex};`};
`

const Top = props => css`
  ${Base(props)};
  bottom: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-bottom: ${props.offset}px;
`

const Bottom = props => css`
  ${Base(props)};
  top: 100%;
  left: 50%;
  transform: translateX(-50%);
  margin-top: ${props.offset}px;
`

const Left = props => css`
  ${Base(props)};
  right: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-right: ${props.offset}px;
`

const Right = props => css`
  ${Base(props)};
  left: 100%;
  top: 50%;
  transform: translateY(-50%);
  margin-left: ${props.offset}px;
`

const BaseToolTop = ({fn, children, ...props}) => (
  <div css={fn(props)}>{children}</div>
)

BaseToolTop.propTypes = {
  fn: PropTypes.func.isRequired,
  children: PropTypes.any.isRequired,
  offset: PropTypes.number,
  open: PropTypes.bool,
  zIndex: PropTypes.number,
  fadeEasing: easingPropType,
  fadeDuration: PropTypes.number,
}

const tooltips = {
  left: ({children, ...props}) => BaseToolTop({fn: Left, children, ...props}),
  top: ({children, ...props}) => BaseToolTop({fn: Top, children, ...props}),
  right: ({children, ...props}) => BaseToolTop({fn: Right, children, ...props}),
  bottom: ({children, ...props}) =>
    BaseToolTop({fn: Bottom, children, ...props}),
}

const Tooltip = ({
  children,
  offset,
  open,
  placement,
  zIndex,
  fadeDuration,
  fadeEasing,
}) => {
  const Component = tooltips[placement] || tooltips.top
  return (
    open && (
      <Component
        offset={offset}
        zIndex={zIndex}
        fadeDuration={fadeDuration}
        fadeEasing={fadeEasing}
      >
        {children}
      </Component>
    )
  )
}

Tooltip.propTypes = {
  children: PropTypes.any.isRequired,
  offset: PropTypes.number,
  open: PropTypes.bool,
  placement: PropTypes.string,
  zIndex: PropTypes.number,
  fadeEasing: easingPropType,
  fadeDuration: PropTypes.number,
}

export default Tooltip
