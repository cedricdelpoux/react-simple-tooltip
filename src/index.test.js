import {mount} from "enzyme"
import React from "react"
import Tooltip from "./index"
import TooltipElement from "./components/Tooltip"

/*eslint-env browser*/

jest.useFakeTimers()

describe("Tooltip", () => {
  const render = props =>
    mount(
      <Tooltip content={"my content"} {...props}>
        <button>Hover Me</button>
      </Tooltip>
    )

  const renderWithoutChildren = props =>
    mount(<Tooltip content={"my content"} {...props} />)

  it("should render", () => {
    const wrapper = render()
    expect(wrapper).toMatchSnapshot()
  })

  it("Should render without children", () => {
    const wrapper = renderWithoutChildren()
    expect(wrapper).toMatchSnapshot()
  })

  it("Should render when there is an offset", () => {
    const arrowSize = 8
    const arrowOffset = 8
    const wrapper = render({arrow: arrowSize, offset: arrowOffset})
    const toolTip = wrapper.find(TooltipElement)
    expect(toolTip.prop("offset")).toEqual(arrowSize + arrowOffset)
  })

  it("should open when user hovers and close when the mouse leaves", () => {
    const wrapper = render()
    const container = wrapper
      .findWhere(n => typeof n.prop("onMouseEnter") === "function")
      .first()
    const toolTip = wrapper.find(TooltipElement)

    // Expect tooltip to be closed by default
    expect(wrapper.state("open")).toEqual(false)
    expect(toolTip.prop("open")).toEqual(false)

    // Simulate a mouse enter event
    container.simulate("mouseEnter")
    wrapper.update()

    // Expect the tooltip to be open
    expect(wrapper.state("open")).toEqual(true)
    expect(toolTip.prop("open")).toEqual(true)

    // Simulate a mouse leave event
    container.simulate("mouseLeave")
    wrapper.update()

    // Expect the tooltip to be closed
    expect(wrapper.state("open")).toEqual(false)
    expect(toolTip.prop("open")).toEqual(false)
  })

  it("should open when user hovers and close when the mouse leaves, after the timeout", () => {
    const exitTimeout = 1000
    const wrapper = render({exitTimeout})
    const container = wrapper
      .findWhere(n => typeof n.prop("onMouseEnter") === "function")
      .first()
    const toolTip = wrapper.find(TooltipElement)

    // Expect tooltip to be closed by default
    expect(wrapper.state("open")).toEqual(false)
    expect(toolTip.prop("open")).toEqual(false)

    // Simulate a mouse enter event
    container.simulate("mouseEnter")
    wrapper.update()

    // Expect the tooltip to be open
    expect(wrapper.state("open")).toEqual(true)
    expect(toolTip.prop("open")).toEqual(true)

    // Simulate a mouse leave event
    container.simulate("mouseLeave")
    wrapper.update()

    // Expect the tooltip not to be closed
    expect(wrapper.state("open")).toEqual(true)
    expect(toolTip.prop("open")).toEqual(true)

    // Expect that the timeout was called once, with the duration
    expect(setTimeout).toHaveBeenCalledTimes(1)
    expect(setTimeout).toHaveBeenLastCalledWith(
      expect.any(Function),
      exitTimeout
    )
  })
})
