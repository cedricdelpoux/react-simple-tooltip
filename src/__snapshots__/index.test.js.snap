// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Should render without children 1`] = `
<Tooltip.Wrapper
  arrow={8}
  background="#000"
  border="#000"
  color="#fff"
  content="my content"
  fadeDuration={0}
  fadeEasing="linear"
  fixed={false}
  fontFamily="inherit"
  fontSize="inherit"
  offset={0}
  padding={16}
  placement="top"
  radius={0}
  zIndex={1}
>
  <EmotionCssPropInternal
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "1oin9j2",
        "next": undefined,
        "styles": "
                
        position: relative;
        display: inline-block;
      ;
                
              ",
      }
    }
  >
    <div
      className="css-1oin9j2"
    >
      <Tooltip
        fadeDuration={0}
        fadeEasing="linear"
        offset={8}
        open={true}
        placement="top"
        zIndex={1}
      >
        <top
          fadeDuration={0}
          fadeEasing="linear"
          offset={8}
          zIndex={1}
        >
          <EmotionCssPropInternal
            __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseToolTop"
            __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
            css={
              Object {
                "map": undefined,
                "name": "iw0mcy",
                "next": undefined,
                "styles": "
                
                position: absolute;
                0;
                z-index: 1;;
              ;
                bottom: 100%;
                left: 50%;
                transform: translateX(-50%);
                margin-bottom: 8px;
              ",
              }
            }
          >
            <div
              className="css-h9nlcb-BaseToolTop"
            >
              <Tooltip.Bubble
                background="#000"
                border="#000"
                color="#fff"
                fontFamily="inherit"
                fontSize="inherit"
                padding={16}
                radius={0}
              >
                <EmotionCssPropInternal
                  __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Bubble"
                  __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                  css={
                    Object {
                      "map": undefined,
                      "name": "2zis3v",
                      "next": undefined,
                      "styles": "
                          color: #fff;
                          background: #000;
                          border-radius: 0;
                          border: 1px solid #000;
                          padding: 16px;
                          font-size: inherit;
                          font-family: inherit;
                        ",
                    }
                  }
                >
                  <div
                    className="css-s1siyn-Bubble"
                  >
                    <Tooltip.Arrow
                      background="#000"
                      border="#000"
                      color="#fff"
                      placement="top"
                      width={8}
                    >
                      <top
                        background="#000"
                        border="#000"
                        width={8}
                      >
                        <EmotionCssPropInternal
                          __EMOTION_LABEL_PLEASE_DO_NOT_USE__="BaseArrow"
                          __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
                          css={
                            Object {
                              "map": undefined,
                              "name": "5qs1ji",
                              "next": undefined,
                              "styles": "
                              
                              position: absolute;
                              width: 8px;
                              height: 8px;
                              background: #000;
                            ;
                              transform: translateX(-50%) translateY(-50%) rotateZ(45deg);
                              top: 100%;
                              left: 50%;
                              border-right: 1px solid #000;
                              border-bottom: 1px solid #000;
                            ",
                            }
                          }
                        >
                          <div
                            className="css-fiwddr-BaseArrow"
                          />
                        </EmotionCssPropInternal>
                      </top>
                    </Tooltip.Arrow>
                    my content
                  </div>
                </EmotionCssPropInternal>
              </Tooltip.Bubble>
            </div>
          </EmotionCssPropInternal>
        </top>
      </Tooltip>
    </div>
  </EmotionCssPropInternal>
</Tooltip.Wrapper>
`;

exports[`Tooltip should render 1`] = `
<Tooltip.Wrapper
  arrow={8}
  background="#000"
  border="#000"
  color="#fff"
  content="my content"
  fadeDuration={0}
  fadeEasing="linear"
  fixed={false}
  fontFamily="inherit"
  fontSize="inherit"
  offset={0}
  padding={16}
  placement="top"
  radius={0}
  zIndex={1}
>
  <EmotionCssPropInternal
    __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
    css={
      Object {
        "map": undefined,
        "name": "1oin9j2",
        "next": undefined,
        "styles": "
                
        position: relative;
        display: inline-block;
      ;
                
              ",
      }
    }
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="css-1oin9j2"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button>
        Hover Me
      </button>
      <Tooltip
        fadeDuration={0}
        fadeEasing="linear"
        offset={8}
        open={false}
        placement="top"
        zIndex={1}
      />
    </div>
  </EmotionCssPropInternal>
</Tooltip.Wrapper>
`;
